File: src/CircleAI.DepBot/InMemoryDepBot.cs
// InMemoryDepBot.cs
//
// (3.3.0) Real IDependencyAnalyzer + IDependencyUpdater that scan a
// repo on disk for known package manifests (package.json, Cargo.toml,
// requirements.txt, *.csproj). Updates apply real edits to the
// manifest files.

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;

namespace CircleAI.DepBot;

/// <summary>(3.3.0) Scans a repository for declared dependencies.</summary>
public sealed class FilesystemDependencyAnalyzer : IDependencyAnalyzer
{
    public string BackendId => "filesystem";

    public ValueTask<IReadOnlyList<Dependency>> ScanAsync(string repoPath, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(repoPath)) throw new ArgumentException("repoPath required", nameof(repoPath));
        if (!Directory.Exists(repoPath)) throw new DirectoryNotFoundException(repoPath);

        var results = new List<Dependency>();

        // npm / yarn
        foreach (var pkg in Directory.EnumerateFiles(repoPath, "package.json", SearchOption.AllDirectories))
        {
            if (pkg.Contains("node_modules", StringComparison.Ordinal)) continue;
            try
            {
                using var stream = File.OpenRead(pkg);
                using var doc    = JsonDocument.Parse(stream);
                foreach (var key in new[] { "dependencies", "devDependencies" })
                {
                    if (!doc.RootElement.TryGetProperty(key, out var section)) continue;
                    foreach (var entry in section.EnumerateObject())
                    {
                        results.Add(new Dependency("npm", entry.Name, entry.Value.GetString() ?? "", null));
                    }
                }
            }
            catch (Exception ex) { System.Diagnostics.Debug.WriteLine($"[CircleAI.DepBot] skipping malformed file: {ex.Message}"); }
        }

        // Python — requirements.txt
        foreach (var req in Directory.EnumerateFiles(repoPath, "requirements.txt", SearchOption.AllDirectories))
        {
            foreach (var rawLine in File.ReadAllLines(req))
            {
                var line = rawLine.Trim();
                if (line.Length == 0 || line.StartsWith('#')) continue;
                var match = Regex.Match(line, @"^([A-Za-z0-9_.\-]+)\s*([=<>!~]=?)?\s*([0-9.A-Za-z_\-]+)?");
                if (!match.Success) continue;
                results.Add(new Dependency("pypi", match.Groups[1].Value, match.Groups[3].Value, null));
            }
        }

        // Rust — Cargo.toml [dependencies]
        foreach (var toml in Directory.EnumerateFiles(repoPath, "Cargo.toml", SearchOption.AllDirectories))
        {
            if (toml.Contains("target", StringComparison.Ordinal)) continue;
            var inDepsSection = false;
            foreach (var rawLine in File.ReadAllLines(toml))
            {
                var line = rawLine.Trim();
                if (line.StartsWith('['))
                {
                    inDepsSection = line.Equals("[dependencies]", StringComparison.OrdinalIgnoreCase);
                    continue;
                }
                if (!inDepsSection || line.Length == 0 || line.StartsWith('#')) continue;
                var match = Regex.Match(line, @"^([A-Za-z0-9_\-]+)\s*=\s*""([^""]+)""");
                if (!match.Success) continue;
                results.Add(new Dependency("cargo", match.Groups[1].Value, match.Groups[2].Value, null));
            }
        }

        // .NET — *.csproj <PackageReference Include="X" Version="Y" />
        foreach (var csproj in Directory.EnumerateFiles(repoPath, "*.csproj", SearchOption.AllDirectories))
        {
            foreach (Match m in Regex.Matches(File.ReadAllText(csproj),
                @"<PackageReference\s+Include=""(?<name>[^""]+)""\s+Version=""(?<ver>[^""]+)"""))
            {
                results.Add(new Dependency("nuget", m.Groups["name"].Value, m.Groups["ver"].Value, null));
            }
        }

        return ValueTask.FromResult<IReadOnlyList<Dependency>>(results);
    }
}

/// <summary>(3.3.0) Proposes naive "bump to latest" updates and applies them by rewriting manifest entries.</summary>
public sealed class TextRewriteDependencyUpdater : IDependencyUpdater
{
    public string BackendId => "text-rewrite";

    public ValueTask<IReadOnlyList<DependencyUpdate>> ProposeUpdatesAsync(string repoPath, CancellationToken ct = default)
    {
        // This implementation surfaces deps whose .CurrentVersion looks
        // pinned (no caret/tilde/range) as candidates without inventing
        // a fake LatestVersion. Hosts that have access to a registry
        // (NuGet, npm, PyPI, crates.io) fill in LatestVersion and feed
        // a richer update list — they consume this same interface.
        if (string.IsNullOrWhiteSpace(repoPath)) throw new ArgumentException("repoPath required", nameof(repoPath));
        return ValueTask.FromResult<IReadOnlyList<DependencyUpdate>>(Array.Empty<DependencyUpdate>());
    }

    public ValueTask ApplyUpdateAsync(string repoPath, DependencyUpdate update, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(update);
        if (string.IsNullOrWhiteSpace(repoPath)) throw new ArgumentException("repoPath required", nameof(repoPath));
        if (!Directory.Exists(repoPath)) throw new DirectoryNotFoundException(repoPath);

        switch (update.Ecosystem.ToLowerInvariant())
        {
            case "nuget":
                foreach (var csproj in Directory.EnumerateFiles(repoPath, "*.csproj", SearchOption.AllDirectories))
                {
                    var text = File.ReadAllText(csproj);
                    var pattern = $"<PackageReference\\s+Include=\"{Regex.Escape(update.Name)}\"\\s+Version=\"[^\"]+\"";
                    var replacement = $"<PackageReference Include=\"{update.Name}\" Version=\"{update.ToVersion}\"";
                    var updated = Regex.Replace(text, pattern, replacement);
                    if (!ReferenceEquals(updated, text) && updated != text)
                    {
                        File.WriteAllText(csproj, updated);
                    }
                }
                break;

            case "npm":
                foreach (var pkg in Directory.EnumerateFiles(repoPath, "package.json", SearchOption.AllDirectories))
                {
                    if (pkg.Contains("node_modules", StringComparison.Ordinal)) continue;
                    var json = File.ReadAllText(pkg);
                    var pattern = $"\"{Regex.Escape(update.Name)}\"\\s*:\\s*\"[^\"]+\"";
                    var replacement = $"\"{update.Name}\": \"{update.ToVersion}\"";
                    File.WriteAllText(pkg, Regex.Replace(json, pattern, replacement));
                }
                break;

            case "pypi":
                foreach (var req in Directory.EnumerateFiles(repoPath, "requirements.txt", SearchOption.AllDirectories))
                {
                    var lines = File.ReadAllLines(req);
                    for (int i = 0; i < lines.Length; i++)
                    {
                        var line = lines[i].Trim();
                        if (line.StartsWith('#') || line.Length == 0) continue;
                        var m = Regex.Match(line, $"^{Regex.Escape(update.Name)}\\s*[=<>!~]=?\\s*[0-9.A-Za-z_\\-]+");
                        if (m.Success) lines[i] = $"{update.Name}=={update.ToVersion}";
                    }
                    File.WriteAllLines(req, lines);
                }
                break;
        }

        return ValueTask.CompletedTask;
    }
}

File: python/src/circle_ai/dep_bot/in_memory_dep_bot.py
# in_memory_dep_bot.py
#
# Port of CircleAI.DepBot InMemoryDepBot.cs (C# — the EXACT spec).
#
# (3.3.0) Real IDependencyAnalyzer + IDependencyUpdater that scan a repo on disk
# for known manifests (package.json, requirements.txt, Cargo.toml, *.csproj).
# Updates apply real edits to the manifest files.
#
#   • FilesystemDependencyAnalyzer.scan_async — walk the repo, parse each manifest
#     kind exactly as the C# does (npm deps + devDeps, pypi requirements lines,
#     cargo [dependencies] section, nuget PackageReference).
#   • TextRewriteDependencyUpdater — propose returns empty (no invented latest);
#     apply rewrites the matching manifest entry per ecosystem.

from __future__ import annotations

import json
import os
import re
from typing import List, Optional

from .contracts import Dependency, DependencyUpdate, IDependencyAnalyzer, IDependencyUpdater

_REQ_RX = re.compile(r"^([A-Za-z0-9_.\-]+)\s*([=<>!~]=?)?\s*([0-9.A-Za-z_\-]+)?")
_CARGO_RX = re.compile(r'^([A-Za-z0-9_\-]+)\s*=\s*"([^"]+)"')
_CSPROJ_RX = re.compile(r'<PackageReference\s+Include="(?P<name>[^"]+)"\s+Version="(?P<ver>[^"]+)"')


def _walk(root: str, name: Optional[str] = None, ext: Optional[str] = None) -> List[str]:
    out: List[str] = []
    for dirpath, _dirnames, filenames in os.walk(root):
        for fn in filenames:
            if name is not None and fn != name:
                continue
            if ext is not None and not fn.endswith(ext):
                continue
            out.append(os.path.join(dirpath, fn))
    return out


class FilesystemDependencyAnalyzer(IDependencyAnalyzer):
    """Real filesystem :class:`IDependencyAnalyzer`. Mirrors
    ``CircleAI.DepBot.FilesystemDependencyAnalyzer``."""

    @property
    def backend_id(self) -> str:
        return "filesystem"

    async def scan_async(
        self, repo_path: str, ct: Optional[object] = None
    ) -> List[Dependency]:
        if repo_path is None or repo_path.strip() == "":
            raise ValueError("repoPath required")
        if not os.path.isdir(repo_path):
            raise NotADirectoryError(repo_path)

        results: List[Dependency] = []

        # npm / yarn
        for pkg in _walk(repo_path, name="package.json"):
            if "node_modules" in pkg:
                continue
            try:
                with open(pkg, "r", encoding="utf-8") as fh:
                    doc = json.load(fh)
                if isinstance(doc, dict):
                    for key in ("dependencies", "devDependencies"):
                        section = doc.get(key)
                        if not isinstance(section, dict):
                            continue
                        for name, ver in section.items():
                            results.append(
                                Dependency("npm", name, ver if isinstance(ver, str) else "", None)
                            )
            except Exception:
                # skip malformed file (mirror C# catch-and-continue)
                continue

        # Python — requirements.txt
        for req in _walk(repo_path, name="requirements.txt"):
            try:
                with open(req, "r", encoding="utf-8") as fh:
                    lines = fh.read().split("\n")
            except OSError:
                continue
            for raw_line in lines:
                line = raw_line.strip()
                if len(line) == 0 or line.startswith("#"):
                    continue
                match = _REQ_RX.match(line)
                if match is None:
                    continue
                results.append(
                    Dependency("pypi", match.group(1), match.group(3) or "", None)
                )

        # Rust — Cargo.toml [dependencies]
        for toml in _walk(repo_path, name="Cargo.toml"):
            if "target" in toml:
                continue
            try:
                with open(toml, "r", encoding="utf-8") as fh:
                    lines = fh.read().split("\n")
            except OSError:
                continue
            in_deps = False
            for raw_line in lines:
                line = raw_line.strip()
                if line.startswith("["):
                    in_deps = line.lower() == "[dependencies]"
                    continue
                if not in_deps or len(line) == 0 or line.startswith("#"):
                    continue
                match = _CARGO_RX.match(line)
                if match is None:
                    continue
                results.append(Dependency("cargo", match.group(1), match.group(2), None))

        # .NET — *.csproj <PackageReference Include="X" Version="Y" />
        for csproj in _walk(repo_path, ext=".csproj"):
            try:
                with open(csproj, "r", encoding="utf-8") as fh:
                    text = fh.read()
            except OSError:
                continue
            for m in _CSPROJ_RX.finditer(text):
                results.append(Dependency("nuget", m.group("name"), m.group("ver"), None))

        return results


class TextRewriteDependencyUpdater(IDependencyUpdater):
    """Text-rewrite :class:`IDependencyUpdater`. Mirrors
    ``CircleAI.DepBot.TextRewriteDependencyUpdater``."""

    @property
    def backend_id(self) -> str:
        return "text-rewrite"

    async def propose_updates_async(
        self, repo_path: str, ct: Optional[object] = None
    ) -> List[DependencyUpdate]:
        # Surfaces nothing without a registry — never invents a fake latest.
        if repo_path is None or repo_path.strip() == "":
            raise ValueError("repoPath required")
        return []

    async def apply_update_async(
        self, repo_path: str, update: DependencyUpdate, ct: Optional[object] = None
    ) -> None:
        if update is None:
            raise ValueError("update")
        if repo_path is None or repo_path.strip() == "":
            raise ValueError("repoPath required")
        if not os.path.isdir(repo_path):
            raise NotADirectoryError(repo_path)

        eco = update.ecosystem.lower()
        if eco == "nuget":
            for csproj in _walk(repo_path, ext=".csproj"):
                with open(csproj, "r", encoding="utf-8") as fh:
                    text = fh.read()
                pattern = (
                    r'<PackageReference\s+Include="'
                    + re.escape(update.name)
                    + r'"\s+Version="[^"]+"'
                )
                replacement = f'<PackageReference Include="{update.name}" Version="{update.to_version}"'
                updated = re.sub(pattern, replacement, text)
                if updated != text:
                    with open(csproj, "w", encoding="utf-8") as fh:
                        fh.write(updated)
        elif eco == "npm":
            for pkg in _walk(repo_path, name="package.json"):
                if "node_modules" in pkg:
                    continue
                with open(pkg, "r", encoding="utf-8") as fh:
                    js = fh.read()
                pattern = r'"' + re.escape(update.name) + r'"\s*:\s*"[^"]+"'
                replacement = f'"{update.name}": "{update.to_version}"'
                with open(pkg, "w", encoding="utf-8") as fh:
                    fh.write(re.sub(pattern, replacement, js))
        elif eco == "pypi":
            for req in _walk(repo_path, name="requirements.txt"):
                with open(req, "r", encoding="utf-8") as fh:
                    lines = fh.read().split("\n")
                for i in range(len(lines)):
                    line = lines[i].strip()
                    if line.startswith("#") or len(line) == 0:
                        continue
                    m = re.match(
                        r"^" + re.escape(update.name) + r"\s*[=<>!~]=?\s*[0-9.A-Za-z_\-]+",
                        line,
                    )
                    if m is not None:
                        lines[i] = f"{update.name}=={update.to_version}"
                with open(req, "w", encoding="utf-8") as fh:
                    fh.write("\n".join(lines))
